### PR TITLE
cross-sample cov

### PR DESCRIPTION
**Pull request by xfangcosmo**
Modified code to incorporate cross lens sample covariance.
Extended the lens and src tomographic bin limit from 10 to 20.
Extended corresponding lens, src combination limits, as well as nz splines.
Fixed a bug in Gaussian clxgl cov shot noise term N23.

### CODE REVIEW

File: cosmolike_core/theory/covariances_real_binned_fullsky_nonlimber_w.c [hunk 1]
```diff
@@ -841,10 +841,16 @@ double func_for_cov_G_cl(double l, int *ar){
...
-  if (n1 == n4){N14= 1./(nlens(n1)*survey.n_gal_conversion_factor);}
-  if (n2 == n3){N23= 1./(nlens(n2)*survey.n_gal_conversion_factor);}
-  if (n2 == n4){N24= 1./(nlens(n2)*survey.n_gal_conversion_factor);}
+  double nlens13, nlens14, nlens23, nlens24;
+  nlens13 = tomo.n_lens_ij[n1][n3];
+  nlens14 = tomo.n_lens_ij[n1][n4];
+  nlens23 = tomo.n_lens_ij[n2][n3];
+  nlens24 = tomo.n_lens_ij[n2][n4];  
+
+  if (nlens13>0.){N13= nlens13/(nlens(n1)*nlens(n3)*survey.n_gal_conversion_factor);}
```
Comment (elikrause) [inline cosmolike_core/theory/covariances_real_binned_fullsky_nonlimber_w.c]: why are these if statements needed?

File: cosmolike_core/theory/covariances_real_binned_fullsky_nonlimber_w.c [hunk 2]
```diff
@@ -859,8 +865,13 @@ double func_for_cov_G_cl_gl(double l, int *ar){
...
   C14 = C_ggl_IA_tab(l,n1,n4);
   C23 = C_cl_tomo_nonlimber_interp(l,n2,n3);
-  if (n1 == n3){N13= 1./(nlens(n1)*survey.n_gal_conversion_factor);}
-  if (n2 == n3){N23= 1./(nlens(n1)*survey.n_gal_conversion_factor);}
+
+  double nlens13, nlens23;
+  nlens13 = tomo.n_lens_ij[n1][n3];
+  nlens23 = tomo.n_lens_ij[n2][n3];
+
+  if (nlens13>0.){N13= nlens13/(nlens(n1)*nlens(n3)*survey.n_gal_conversion_factor);}
```
Comment (elikrause) [inline cosmolike_core/theory/covariances_real_binned_fullsky_nonlimber_w.c]: (same as above, needed if tomo.n_lens_ij[i][j] is properly initialized?)

File: cosmolike_core/theory/structs.c [hunk 3]
```diff
@@ -63,18 +63,19 @@ cosmopara cosmology = {.A_s = 0., .sigma_8=0., .alpha_s =0.0, .M_nu =0., .Omega_
...
+  double shear_zmin[20];
+  double n_source[20];
   int clustering_Nbin; // number of tomography bins
   int clustering_Npowerspectra;// number of tomography power spectra+2+3+...+Nbin
-  double clustering_zmax[10];
-  double clustering_zmin[10];
-  double n_lens[10];
+  double clustering_zmax[20];
+  double clustering_zmin[20];
+  double n_lens[20];
```
Comment (elikrause) [inline cosmolike_core/theory/structs.c]: why keep n_lens, if new code uses only n_lens_ij?

**Review by elikrause — COMMENTED** [review]
Didn't run it yet - do you have an example config file using n_lens_ij?

**Comment by xfangcosmo** [comment]
@elikrause Edited, also fixed for flat sky

**Pull request merged**